cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -96,6 +100,31 @@ public void upgrade() throws Exception {
 
     // Upgrade the datasets instance meta table
     datasetInstanceMDSUpgrader.upgrade();
+
+    // upgrade all the filesets base paths
+    for (DatasetSpecification fileSetSpec : datasetInstanceMDSUpgrader.getFileSetsSpecs()) {
+      upgradeFileSet(fileSetSpec);
+    }
+
+  }
+
+  /**
+   * Upgrades the {@link FileSet} and also its embedded filesets if any by moving the base path under
+   * namespaced directory
+   *
+   * @param dsSpec the {@link DatasetSpecification} of the {@link FileSet} to be upgraded
+   * @throws IOException
+   */
+  private void upgradeFileSet(DatasetSpecification dsSpec) throws IOException {
+    for (DatasetSpecification embeddedDsSpec : dsSpec.getSpecifications().values()) {
+      upgradeFileSet(embeddedDsSpec);
+    }
+    String basePath = FileSetProperties.getBasePath(dsSpec.getProperties());
+    Location oldLocation = locationFactory.create(basePath);
+    Location newlocation = locationFactory.create(Constants.DEFAULT_NAMESPACE)
+      .append(cConf.get(Constants.Dataset.DATA_DIR)).append(basePath);
+    LOG.info("Upgrading base path for dataset {} from {} to {}", dsSpec.getName(), oldLocation, newlocation);
+    renameLocation(oldLocation, newlocation);
```

Comment:
No we use rename here: 
https://github.com/caskdata/cdap/blob/fa44c400e335b13e00a495f1f7ee91af3c3e091c/cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrader.java#L57

So, if the basePath got moved initially then for other datasets it will say that it might have been moved and will not throw does FileNotFoundException as the renameTo does as we skip rename if it doesn't. So, I think that should be fine.


But I haven't tested the FileSet upgrade for FileSets which have same basepath. That is the only thing which I have on my plate for today. I have tested it with FileSet example.